src/poh_recorder.rs

```diff
@@ -137,32 +152,98 @@ mod tests {
         let h1 = hash(b"hello world!");
         let tx = test_tx();
-        poh_recorder
-            .record(h1, vec![tx.clone()], &entry_sender)
-            .unwrap();
+        poh_recorder.record(h1, vec![tx.clone()], &leader).unwrap();
         //get some events
         let e = entry_receiver.recv().unwrap();
         assert_eq!(e[0].tick_height, 0); // super weird case, but ok!
 
-        poh_recorder.tick(&bank, &entry_sender).unwrap();
+        poh_recorder.tick(&leader).unwrap();
         let e = entry_receiver.recv().unwrap();
         assert_eq!(e[0].tick_height, 1);
 
-        poh_recorder.tick(&bank, &entry_sender).unwrap();
+        poh_recorder.tick(&leader).unwrap();
         let e = entry_receiver.recv().unwrap();
         assert_eq!(e[0].tick_height, 2);
 
         // max tick height reached
-        assert!(poh_recorder.tick(&bank, &entry_sender).is_err());
-        assert!(poh_recorder.record(h1, vec![tx], &entry_sender).is_err());
+        assert!(poh_recorder.tick(&leader).is_err());
+        assert!(poh_recorder.record(h1, vec![tx], &leader).is_err());
 
         //make sure it handles channel close correctly
         drop(entry_receiver);
-        assert!(poh_recorder.tick(&bank, &entry_sender).is_err());
+        assert!(poh_recorder.tick(&leader).is_err());
+    }
+
+    #[test]
+    fn test_poh_recorder_tick_cache() {
+        let (genesis_block, _mint_keypair) = GenesisBlock::new(2);
+        let bank = Arc::new(Bank::new(&genesis_block));
+        let prev_id = bank.last_id();
+        let (entry_sender, entry_receiver) = channel();
+        let poh_recorder = PohRecorder::new(0, prev_id);
+
+        let leader = WorkingLeader {
+            bank,
+            sender: entry_sender,
+            min_tick_height: 1,
+            max_tick_height: 2,
+        };
+
+        // tick should be cached
+        assert!(poh_recorder.tick(&leader).is_err());
+        assert!(entry_receiver.try_recv().is_err());
+
+        // leader should be at the right height
+        poh_recorder.tick(&leader).unwrap();
+
+        let e = entry_receiver.recv().unwrap();
+        assert_eq!(e[0].tick_height, 1);
+        let e = entry_receiver.recv().unwrap();
+        assert_eq!(e[0].tick_height, 2);
+    }
+
+    #[test]
+    fn test_poh_recorder_tick_cache_old_working_leader() {
+        let (genesis_block, _mint_keypair) = GenesisBlock::new(2);
+        let bank = Arc::new(Bank::new(&genesis_block));
+        let prev_id = bank.last_id();
+        let (entry_sender, entry_receiver) = channel();
+        let poh_recorder = PohRecorder::new(0, prev_id);
+
+        let leader = WorkingLeader {
+            bank,
+            sender: entry_sender,
+            min_tick_height: 1,
+            max_tick_height: 1,
+        };
+
+        // tick should be cached
+        assert_matches!(
+            poh_recorder.tick(&leader),
+            Err(Error::PohRecorderError(
+                PohRecorderError::MinHeightNotReached
```

Comment:
Are you certain we need this case? If the working bank is never used by anyone but the TPU, can we send those ticks and let that bank fork be the cache?

Reply:
@garious that error can be seen If record is called before PoH reaches the minimum tick height. Record should probably assert that case because the only reason why that would happen is due to a programmer error.

Reply:
Interesting. We should sync up. I thought pre-MinHeight ticks were what we were calling "virtual ticks". If so, it's not a program error to write them to the working bank. If the TVU votes before MinHeight, it should discard that WorkingBank and reset PoH by sending it a new one.

Reply:
@garious, that’s a bit more complicated then I was going for initially.   It would be nice to optimistically start generating a new bank, but for that to work we would need to hold of broadcasting any blobs.  Otherwise there are two ledgers broadcast by the leader.  

Right now I am assuming that when banking stage start calling record, the TVU is no longer able to reset PoHRecorder.  So reset would error out.  So letting the TPU register a WorkingBank is a commitment to a specific PoH and to accept or reject the previous slot.